analyzers/src/SonarAnalyzer.CSharp/Rules/ExceptionShouldNotBeThrownFromUnexpectedMethods.cs

```diff
@@ -85,61 +83,83 @@ protected override void Initialize(SonarAnalysisContext context)
+            if (isTrackedSyntax(syntax))
             {
-                return;
+                ReportOnInvalidThrowStatement(analysisContext, syntax, allowedThrowTypes);
             }
+        }
+
+        private static bool IsTrackedOperator(OperatorDeclarationSyntax declaration) =>
+            TrackedOperators.Contains(declaration.OperatorToken.Kind());
+
+        private static bool IsTrackedMethod(MethodDeclarationSyntax declaration, SemanticModel semanticModel) =>
+            HasTrackedMethodOrAttributeName(declaration)
+            && semanticModel.GetDeclaredSymbol(declaration) is { } methodSymbol
+            && HasTrackedMethodOrAttributeType(methodSymbol);
 
-            ReportOnInvalidThrowStatement(analysisContext, syntax, allowedThrowTypes);
+        private static bool HasTrackedMethodOrAttributeName(MethodDeclarationSyntax declaration)
+        {
+            var name = declaration.Identifier.ValueText;
+            return name == "Equals"
+                || name == "GetHashCode"
+                || name == "ToString"
+                || name == "Dispose"
+                || name == "Equals"
+                || declaration.AttributeLists.SelectMany(list => list.Attributes).Any(x => x.ArgumentList == null && x.Name.ToStringContains("ModuleInitializer"));
         }
 
-        private void ReportOnInvalidThrowStatement(SyntaxNodeAnalysisContext analysisContext,
+        private static bool HasTrackedMethodOrAttributeType(IMethodSymbol methodSymbol) =>
+            methodSymbol.IsObjectEquals()
+            || methodSymbol.IsObjectGetHashCode()
+            || methodSymbol.IsObjectToString()
+            || methodSymbol.IsIDisposableDispose()
+            || methodSymbol.IsIEquatableEquals()
+            || IsModuleInitializer(methodSymbol);
+
+        private static bool IsModuleInitializer(IMethodSymbol methodSymbol) =>
+            methodSymbol.AnyAttributeDerivesFrom(KnownType.System_Runtime_CompilerServices_ModuleInitializerAttribute);
+
+        private static void ReportOnInvalidThrowStatement(SyntaxNodeAnalysisContext analysisContext,
             SyntaxNode node, ImmutableArray<KnownType> allowedTypes)
         {
-            var throwToReportOn = node.DescendantNodes()
-                .OfType<ThrowStatementSyntax>()
-                .Where(tss => tss.Expression != null)
-                .Select(tss => new NodeAndSymbol(tss, analysisContext.SemanticModel.GetSymbolInfo(tss.Expression).Symbol))
-                .FirstOrDefault(tuple => tuple.Symbol != null &&
-                    !tuple.Symbol.ContainingType.IsAny(allowedTypes) &&
-                    !tuple.Symbol.ContainingType.DerivesFromAny(allowedTypes))
-                ?.Node;
-
-            if (throwToReportOn != null)
+            if (ExpressionBody(node) is { } expressionBody
+                && ThrowExpressionSyntaxWrapper.IsInstance(expressionBody.Expression)
+                && (ThrowExpressionSyntaxWrapper)expressionBody.Expression is { } throwExpression
```

Comment:
Since `ThrowExpressionSyntaxWrapper` is struct it can never be null. I suggest to use `(ThrowExpressionSyntaxWrapper)expressionBody.Expression is var throwExpression` instead.

Strange thing: the coverage is 100% on this line although the condition (`is { }`) can never be false.